drivers: gps: nrf9160_gps: update GPS driver to use GNSS API

tokangas opened the pull request:
Updated GPS driver to use the new function based GNSS API.

Signed-off-by: Tommi Kangas <tommi.kangas@nordicsemi.no>

NordicBuilder commented:
You can find the documentation preview for this PR at this <a href='http://developer.nordicsemi.com/nRF_Connect_SDK_dev/doc/PR-5233
'>link</a>. It will be updated about 10 minutes after the documentation build succeeds.

*Note: This comment is automatically posted by the Documentation Publishing GitHub Action.*

rlubos reviewed (approved):
LGTM

tokangas commented:
Updated `gps_start()` description in `gps.h`. I've changed how `gps_start()` behaves if GPS is already running when the function is called. Earlier, GPS was restarted, but now the function just returns an error. I don't see why restart functionality would be needed and it would have required additional implementation. GPS can anyway be restarted by calling `gps_stop()` and `gps_start()`.

VeijoPesonen commented on include/drivers/gps.h:
```diff
@@ -329,8 +329,7 @@ struct gps_driver_api {
  *
- * If gps is already running a call to this function will
- * restart the gps.
+ * If gps is already running a call to this function will fail.
```
Return values should be documented.

tokangas replied:
Added @return descriptions for `gps_start()` and `gps_stop()`. I didn't list possible return values as it hasn't been done for any of the functions and I think we should encourage moving to use the new GNSS API anyway.

VeijoPesonen commented on drivers/gps/nrf9160_gps/nrf9160_gps.c:
```diff
@@ -909,37 +606,31 @@ static void timeout_work_fn(struct k_work *work)
 	struct gps_event evt = {
```
There is an idiom for obtaining the data which takes into consideration the delayable nature of the work queue item. Please see the [Delayable Work](
https://docs.zephyrproject.org/latest/reference/kernel/threads/workqueue.html?highlight=container_of#delayable-work) API documentation.

tokangas replied:
I'll fix this.

VeijoPesonen commented on drivers/gps/nrf9160_gps/nrf9160_gps.c:
```diff
@@ -204,173 +237,56 @@ static void notify_event(const struct device *dev, struct gps_event *evt)
-					"initialized state");
-				goto wait;
+			if (pvt_data->flags & NRF_MODEM_GNSS_PVT_FLAG_FIX_VALID) {
+				k_work_cancel_delayable(&drv_data->timeout_work);
```
Would be better to call _k_work_cancel_delayable_sync()_ instead as it was done earlier. This is a executed on the context of the GPS thread so no need to worry about stalling an event queue.

tokangas replied:
Can you explain why `k_work_cancel_delayable_sync()` should be used here? The only difference seems to be that the cancellation is performed synchronously, i.e. the calling thread waits until the delayable work item has been cancelled. I don't see the reason for that here.

VeijoPesonen replied:
> A large number of problems with Zephyr's work queue result from the decision to support cancelling work items that have been submitted, while failing to account for cases where the cancellation could not be performed immediately. It's years too late to change that, but I strongly suspect that a better design would have justified cancelling a work item only in a situation where a subsystem that uses the work queue is being shut down. In that case, you want to cancel, and confirm cancellation, to be sure it's safe to release any resources in the subsystem---or that it's safe to restart it. This is what synchronous cancellation is for.
> 
> It comes down to a question: If you cancel a work item, do you mean to cancel all past submissions/schedules of that item, or do you want to force the work item to an idle state where its handler will not be invoked again? These are very different goals. For the latter, resubmission of a still-running item must be rejected otherwise it may never become idle. For the former, you could probably accomplish the same thing by recording additional state specific to the work item and the events that cause it to be scheduled/submitted.

https://github.com/zephyrproject-rtos/zephyr/issues/35255#issuecomment-840877843

VeijoPesonen replied:
One more thing to consider is that you can't resubmit a work queue item as long as there is one in cancelled state.

````
prevents resubmission of a item being canceled until cancellation completes;
````
https://github.com/zephyrproject-rtos/zephyr/pull/29618#issue-511682580

tokangas replied:
I know there have been problems related to cancelling a work item. I'm just trying to understand the actual issue and how it affects the GPS driver.

My understanding is, that synchronous cancel only blocks when the delayed work item is running (i.e. being executed). In this case it would block until the item stops running. GPS driver uses two delayed works, one sends a timeout event and the second sends one RPC message. Thus, it's unlikely that the execution takes a long time. Work items are (re)scheduled only when a new PVT notification is received from GNSS, i.e. this can happen only once a second. For the synchronous cancel to make any difference, the work being cancelled would have to run for more than one second. Otherwise the cancellation would be completed before the next PVT notification is received anyway.

Also, it seems that scheduling a work item being cancelled is not a problem, submitting it into a work queue is. The code never submits the works directly, but schedules them for execution 5 or 10 seconds later.

https://docs.zephyrproject.org/latest/reference/kernel/threads/workqueue.html#c.k_work_cancel_delayable

_Canceling delayable work does not prevent rescheduling it. It does prevent submitting it until the cancellation completes._

Finally, I don't want the GPS driver thread to be blocked for one second or more. If the thread would be blocked for this long time, the event queue is likely to become full. This would result in PVTs or NMEAs not being processed at all, which would actually be worse than those delayed works not being run as expected.

VeijoPesonen replied:
Ok, sounds good.

VeijoPesonen commented on drivers/gps/nrf9160_gps/nrf9160_gps.c:
```diff
@@ -389,58 +305,54 @@ static void gps_thread(int dev_ptr)
-				k_work_cancel_delayable_sync(
-					&drv_data->blocked_work,
-					&drv_data->work_sync);
+				k_work_cancel_delayable(&drv_data->blocked_work);
```
Would be better to call k_work_cancel_delayable_sync() instead as it was done earlier. This is a executed on the context of the GPS thread so no need to worry about stalling an event queue.

tokangas replied:
Same question here.

VeijoPesonen reviewed (changes requested):
There were couple of things which would need to be fixed.

tokangas commented:
> Since these are obviously false positives, I think we could simply ask @carlescufi to override the CI check once the PR is ready. I don't think there's a way to suppress particular checkpatch warning (other than actually renaming the functions to shut it up).

Although these were false positives, I rephrased the strings a bit to get rid of the warnings.

VeijoPesonen reviewed (approved):
Looks good to me.

tokangas commented:
There's a problem with this PR and asset_tracker_v2 which needs to be sorted out.

tokangas commented:
> There's a problem with this PR and asset_tracker_v2 which needs to be sorted out.

asset_tracker_v2 doesn't call `gps_stop()` before calling `gps_start()` again. I though that this functionality wasn't needed and had removed that, so asset_tracker_v2 failed to start the GPS again after the first fix. I have now implemented support for GPS restart and restored the API to how it was earlier.

Also fixed NMEA string copying so that also NULL termination is copied into the event buffer, although it's not included in the NMEA length.

tokangas commented:
Pushed one more change to correct debug log fix time calculation in cases where periodic mode is not used.

jtguggedal commented on drivers/gps/nrf9160_gps/nrf9160_gps.c:
```diff
@@ -964,36 +655,193 @@ static int init(const struct device *dev, gps_event_handler_t handler)
+		return -EIO;
+	}
+
+	err = nrf_modem_gnss_nmea_mask_set(drv_data->config.nmea_mask);
```
Setting of NMEA mask seems to happen twice

tokangas replied:
Oops, removed the extra setting.

jtguggedal reviewed (commented):
One comment on duplicate code, otherwise looks good!

simensrostad commented on drivers/gps/nrf9160_gps/nrf9160_gps.c:
```diff
@@ -13,87 +13,140 @@
+
+	err = k_msgq_put(&event_msgq, &event, K_NO_WAIT);
+	if (err) {
+		/* Failed to put event into queue. */
```
If this API fails it might be due to the message queue being filled which would indicate that the processing thread might be blocked somehow. The application should probably be notified with an error and/or the message queue should be purged.

tokangas replied:
Are you thinking about a warning/error trace or something else? I think a trace would be nice, but gnss_event_handler() is called from an ISR, so I'm not sure we want to use Logger? If deferred logging is not used, it might cause problems. Would printk() be better?

simensrostad replied:
I just wonder if we can afford not processing events received via the GNSS API when they indirectly translate to some state changing event in the application. (GPS_EVT_START, FIX, TIMEOUT). If we can get into an undefined state in the application and that we should consider a filled message queue in this situation a fatal error really. `k_msgq_put` should never really fail and if it does we should detect it as early as possible. Maybe a run-time assert is the way to go? Or callback to the application (GPS_EVT_ERROR with an error cause). So that we atleast trasfer some awareness of the problem that occurred.

tokangas replied:
I get your point. An assert sounds like a good idea to me, that makes the issue visible immediately. If the message queue becomes full, the application might for example miss a fix, so the code should be changed so that it can process all events fast enough.

tokangas replied:
As discussed offline, I'll add both an assert and an error event in case asserts are disabled.

tokangas replied:
Assert caused an assert loop, probably because the handler is being run in ISR. Because of this I only added sending of GPS_EVT_ERROR event. I also changed the implementation in asset_tracker_v2 so that GPS_EVT_ERROR is handled as a fatal error.

tokangas commented on drivers/gps/nrf9160_gps/nrf9160_gps.c:
```diff
@@ -13,87 +13,140 @@
+static void gnss_event_handler(int event_id)
+{
+	int err = 0;
+	struct event_item event;
```
Changed.

tokangas commented on drivers/gps/nrf9160_gps/nrf9160_gps.c:
```diff
@@ -13,87 +13,140 @@
+	int err = 0;
+	struct event_item event;
+
+	event.id = event_id;
```
Changed.

tokangas commented on drivers/gps/nrf9160_gps/nrf9160_gps.c:
```diff
@@ -13,87 +13,140 @@
+	struct event_item event;
+
+	event.id = event_id;
+	event.data = NULL;
```
Changed.

simensrostad reviewed (approved):
Thanks!
